File: LinearTwoClassClassifier.py
import numpy as np
from data_process import get_data_set
import math
import matplotlib.pyplot as plt


def sigmoid(t):
    s = 1 / (1 + math.exp(-t))
    if s == 0:
        return 0.0001
    elif s == 1:
        return 0.9999
    else:
        return s


def cross_entropy(w, x, y):
    p = y.shape[0]
    xw = np.dot(x, w)
    N = xw.shape[1]
    for i in range(xw.shape[0]):
        for j in range(xw.shape[1]):
            xw[i][j] = sigmoid(xw[i][j])
    cost = np.zeros([1, N])
    for i in range(N):
        for j in range(p):
            cost[0][i] += - y[j][0] * math.log(xw[j][i]) - (1 - y[j][0]) * math.log(1 - xw[j][i])
        cost[0][i] = cost[0][i] / p
    return cost


def approx_grad(w, x, y, delta):
    N = w.shape[0]
    dw = np.identity(N) * delta
    grad = (cross_entropy(w + dw, x, y) - cross_entropy(w, x, y)) / delta
    return grad


def gradient_descent(w, x, y, alpha, delta, n_iter):
    trained_w = w
    cost = 1
    k = 0
    # plotting cost function
    plt.ion()
    plt.figure(1)

    while k < n_iter or cost < 0.01:
        grad = approx_grad(trained_w, x, y, delta)
        trained_w = trained_w - alpha * grad.T / np.linalg.norm(grad)
        cost = cross_entropy(trained_w, x, y)

        print('iteration: ', k, ', cost: ', np.squeeze(cost))
        plt.plot(k, np.squeeze(cost), c='r', marker='.', mec='r')

        k = k + 1

    plt.title('Cost Function Curve in Training')
    plt.xlabel('Iteration')
    plt.ylabel('Cost Value')
    plt.show()
    return trained_w


def plot_Matrix(cm, classes, title='Confusion Matrix in Testing', cmap=plt.cm.Blues):
    plt.rc('font', family='Times New Roman', size='8')  # 设置字体样式、大小

    # 按行进行归一化
    cm = cm.astype('float') / cm.sum(axis=1)[:, np.newaxis]
    # print("Normalized confusion matrix")
    str_cm = cm.astype(np.str).tolist()
    for row in str_cm:
        print('\t'.join(row))
    # 占比1%以下的单元格，设为0，防止在最后的颜色中体现出来
    for i in range(cm.shape[0]):
        for j in range(cm.shape[1]):
            if int(cm[i, j] * 100 + 0.5) == 0:
                cm[i, j] = 0

    fig, ax = plt.subplots()
    im = ax.imshow(cm, interpolation='nearest', cmap=cmap)
    # ax.figure.colorbar(im, ax=ax) # 侧边的颜色条带

    ax.set(xticks=np.arange(cm.shape[1]),
           yticks=np.arange(cm.shape[0]),
           xticklabels=classes, yticklabels=classes,
           title=title,
           ylabel='Actual',
           xlabel='Predicted')

    # 通过绘制格网，模拟每个单元格的边框
    ax.set_xticks(np.arange(cm.shape[1] + 1) - .5, minor=True)
    ax.set_yticks(np.arange(cm.shape[0] + 1) - .5, minor=True)
    ax.grid(which="minor", color="gray", linestyle='-', linewidth=0.2)
    ax.tick_params(which="minor", bottom=False, left=False)

    # 将x轴上的lables旋转45度
    plt.setp(ax.get_xticklabels(), rotation=45, ha="right",
             rotation_mode="anchor")

    # 标注百分比信息
    fmt = 'd'
    thresh = cm.max() / 2.
    for i in range(cm.shape[0]):
        for j in range(cm.shape[1]):
            if int(cm[i, j] * 100 + 0.5) > 0:
                ax.text(j, i, format(int(cm[i, j] * 100 + 0.5), fmt) + '%',
                        ha="center", va="center",
                        color="white" if cm[i, j] > thresh else "black")
    fig.tight_layout()
    plt.savefig('cm.jpg', dpi=300)
    plt.show()


# Train
train_X, train_Y = get_data_set(0)
w0 = np.ones([30, 1])
w0[:, :] = 0

bias_x = np.ones([train_X.shape[0], 1])
train_X = np.column_stack((bias_x, train_X))
new_w = gradient_descent(w0, train_X, train_Y, 0.001, 0.00001, 40000)
print(new_w)

# Test
test_X, test_Y = get_data_set(1)
bias_x_test = np.ones([test_X.shape[0], 1])
test_X = np.column_stack((bias_x_test, test_X))
test_res = np.dot(test_X, new_w)
for i in range(test_res.shape[0]):
    test_res[i][0] = sigmoid(test_res[i][0])
    if test_res[i][0] > 0.5:
        test_res[i][0] = 1
    else:
        test_res[i][0] = 0

# Test Result Metrics
test_P = test_res.shape[0]
total_0 = 0
total_1 = 0
test_0 = 0
test_1 = 0
acc_0 = 0
acc_1 = 0
for i in range(test_P):
    if test_Y[i][0] == 1:
        total_1 = total_1 + 1
        if test_res[i][0] == test_Y[i][0]:
            test_1 = test_1 + 1
    else:
        total_0 = total_0 + 1
        if test_res[i][0] == test_Y[i][0]:
            test_0 = test_0 + 1

acc_0 = test_0 / total_0
acc_1 = test_1 / total_1
acc = (acc_0 + acc_1) / 2
print('test accuracy:', acc)
# plot Confusion Matrix
cm = np.zeros((2, 2))
cm[0][0] = test_0
cm[0][1] = total_0 - test_0
cm[1][0] = total_1 - test_1
cm[1][1] = test_1
plot_Matrix(cm, [0, 1])

File: Linear_TwoClass_Classifier.m
clear
clc
% Load Training data
Mat = csvread('Otonabee_Fluoranthene_training.csv');

% Convert the first column to two class labels 0 & 1(Fluoranthene concentration < or ≥ 0.5)
y = double(logical(Mat(:,1)'>= 0.5));

% Preprocess Training Data
%... Find the average of 111 emission wavelengths caused by each excitation wavelength of each sample
%... Mean Matrix has 29 rows(no. of features) and 336 columns(no. of samples) 
i = 1;
while i <30
Mean(i,:) = mean(Mat(:,(2+111*(i-1)):(1+111*i)),2)'; 
i = i+1;
end
x = Mean;

% Initialize w 
w0 = [4;0;1;-2;-1;-5;2;-4;-3;1;2;-5;6;3;7;-1;8;-2;4;-1;3;-7;2;4;-5;3;-6;-2;2;1]; 
f = @(w)cross_entropy(w,x,y);

%%% Train classifier %%%
[W,fW] = gradient_descent(f,w0,0.01,10000);
figure
plot(fW,'linewidth',1)
axis([-200 11000 0 33])
title('Cost Function Curve in Training')
xlabel('Iteration');ylabel('Cost Value')

%%% Test classifier %%%
% Load test data
Mat1 = csvread('Otonabee_Fluoranthene_test.csv');
% Preprocess test Data
i = 1;
while i <30
Mean1(i,:) = mean(Mat1(:,(2+111*(i-1)):(1+111*i)),2)'; 
i = i+1;
end
x = Mean1;
% actual labels
y_actual = double(logical(Mat1(:,1)'>= 0.5));
P = length(y_actual);
xbar = [ones(1,P); x];
Sig = sigmoid(xbar'*W(:,end));
y_predicted = double(logical(Sig(:,1)'>= 0.5));
% Number of errors
Errors = sum(logical(y_actual~=y_predicted))
% Accuracy of classifier
Accuracy = sum(logical(y_actual==y_predicted))/P
% Confusion Matrix
a = sum((y_actual==1)&(y_predicted==1));
b = sum((y_actual==1)&(y_predicted==0));
c = sum((y_actual==0)&(y_predicted==1));
d = sum((y_actual==0)&(y_predicted==0));
A_balanced = (a/(a+c) + d/(b+d))/2
CM = [a b;c d];
figure
imagesc(CM) % Display image with scaled colors
colorbar
title('Confusion Matrix')
set(gca, 'XTick', [1 2]) 
set(gca,'XTickLabel',{'1','0'}) 
set(gca, 'YTick', [1 2]) 
set(gca,'YTickLabel',{'1','0'}) 
ylabel('Actual Label')
xlabel('Predicted Label')
text(0.75,1,'68 (95.8%)','FontSize',14)
text(1.75,1,'3 (4.2%)','FontSize',14)
text(0.75,2,'3 (2.3%)','FontSize',14)
text(1.75,2,'126 (97.7%)','FontSize',14)
annotation('textbox',[.01 0 .25 .1],'String','A balanced: 96.72%  A: 97% ','EdgeColor','none')

function cost = cross_entropy(w,x,y)
P = length(y);
xbar = [ones(1,P); x];
SIG = sigmoid(xbar'*w);
cost = (1/P) * ...
    sum( ...
    - repmat(y',1,size(w,2)) .* log(SIG)...
    - (1-repmat(y',1,size(w,2))) .* log(1-SIG));
end

function s = sigmoid(x)
s = 1./(1+exp(-x));
s(s==1) = .9999;
s(s==0) = .0001;
end

function [W,fW] = gradient_descent(f,w0,alpha,n_iter)
k = 1;
W = w0;
fW = f(w0);
while k < n_iter
    grad = approx_grad(f,W(:,k),0.0001);
    W(:,k+1) = W(:,k) - alpha*(grad')/norm(grad);
    fW(k+1) = f(W(:,k+1));
    k = k+1;
end
end

function grad = approx_grad(f,w0,delta)
N = length(w0);
dw = delta*eye(N);
grad = ( f(w0+dw) - f(w0) )/delta;
end